jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala

```diff
@@ -267,4 +268,14 @@ class XGBoostGeneralSuite extends FunSuite with PerTest {
 
     assert(booster != null)
   }
+
+  test("repartitionForTrainingGroup with empty partition") {
+    val trainingRDD = sc.parallelize(Ranking.train, 5).mapPartitions(it => {
+      // make one partition empty for testing
+      it.filter(_ => TaskContext.getPartitionId() != 3)
+    })
+    val traingGroupsRDD = XGBoost.repartitionForTrainingGroup(trainingRDD, 4)
```

Comment:
can you call through trainDistributed API, which is the unified interface for XGBClassifier/Regressor 

and you may want to change repartitionForTrainingGroup to private

Reply:
changed the test to call trainDistributed API.

repartitionForTrainingGroup is package level private already just like the repartitionForTraining method.
API user have no access to it, but it is needed to write unit test.

Reply:
as the current unit test is kind of coarse-grained, i.e. we didn't test the most nested call of the function, but test everything along the calling path by calling trainDistributed()

I just noticed that the previous ranking trainer test are also calling repartitionForTrainingGroup()

so you should either change everything to call trainDistributed() and make repartitionForTrainingGroup as private, 

or you should have two unit test cases for everything you want to test, by calling trainDistributed and repartitionForTrainingGroup respecitively

Reply:
> as the current unit test is kind of coarse-grained, i.e. we didn't test the most nested call of the function, but test everything along the calling path by calling trainDistributed()
> 
> I just noticed that the previous ranking trainer test are also calling repartitionForTrainingGroup()
> 
> so you should either change everything to call trainDistributed() and make repartitionForTrainingGroup as private,
> 
> or you should have two unit test cases for everything you want to test, by calling trainDistributed and repartitionForTrainingGroup respectively

1. There is another unit test calling trainDistributed(): test("distributed training with group data")
2. repartitionForTrainingGroup was defined as 'private[spark]' which follows the convention of the existing method 'private[spark] def repartitionForTraining()'
3. If you take a close look of 'test("repartitionForTrainingGroup with group data")', there is no way to verify whether the grouped data is correct or not by calling  trainDistributed() instead of repartitionForTrainingGroup()

Reply:
the existence of this bug shows that test("distributed training with group data") doesn't cover the test with data containing empty partition, 

test("repartitionForTrainingGroup with group data")' should be changed to something like test("distributed training with group data containing empty partition")' 

because we have to ensure that  the code path (and the future change on them) in trainDistributed() before  `repartitionForTrainingGroup` has no impact on the things you are testing here... trainDistributed() is actually the entry for everything outside of this file 


if you intentionally want to test the data integrity produced by  `repartitionForTrainingGroup` keep it as private[spark] is fine and you may want to add back your original test in addition to the current one

Reply:
add the original test back.

1. test("repartitionForTrainingGroup with group data"): test data integrity after re-grouping.
2. test("repartitionForTrainingGroup with group data which has empty partition"): test corner case with empty partition.
3. test("distributed training with group data which has empty partition"): test from trainDistributed() which is the API to the outside.